src/lib.rs

```diff
@@ -427,83 +427,25 @@ impl<'a> RRule<'a> {
-                        next_date = add_month_to_date(next_date);
-                    }
-                } else if start_day == by_month_day_u32 {
-                    if next_date.gt(&start_date) {
-                        next_date = add_month_to_date(next_date);
-                        next_date = next_date.with_day(by_month_day_u32).unwrap();
-                        for _i in 0..interval {
-                            next_date = add_month_to_date(next_date);
-                        }
-                    } else {
-                        next_date = next_date.with_day(by_month_day_u32).unwrap();
-                        for _i in 0..interval {
-                            next_date = add_month_to_date(next_date);
-                        }
-                    }
-                }
-            }
-        } else {
-            loop {
-                if by_month_day.is_empty() {
-                    for _i in 0..interval {
-                        next_date = add_month_to_date(next_date);
-                    }
-                } else {
-                    let by_month_day_u32 = by_month_day.parse::<u32>().unwrap();
-                    let start_day = start_date.day();
-                    if start_day < by_month_day_u32 {
-                        next_date = next_date.with_day(by_month_day_u32).unwrap();
-                        for _i in 0..interval {
-                            next_date = add_month_to_date(next_date);
-                        }
-                    } else if start_day > by_month_day_u32 {
-                        // move forward a month
-                        next_date = add_month_to_date(next_date);
-                        next_date = next_date.with_day(by_month_day_u32).unwrap();
-                        for _i in 0..interval {
-                            next_date = add_month_to_date(next_date);
-                        }
-                    } else if start_day == by_month_day_u32 {
-                        if next_date.ge(&start_date) {
-                            next_date = add_month_to_date(next_date);
-                            next_date = next_date.with_day(by_month_day_u32).unwrap();
-                            for _i in 0..interval {
-                                next_date = add_month_to_date(next_date);
-                            }
-                        } else {
-                            next_date = next_date.with_day(by_month_day_u32).unwrap();
-                            for _i in 0..interval {
-                                next_date = add_month_to_date(next_date);
-                            }
-                        }
-                    }
-                }
-                if next_date.month().eq(&(by_month.parse::<u32>().unwrap())) {
-                    break;
-                }
+        if !by_month.is_empty() {
+            let by_month_u32 = by_month.parse::<u32>().unwrap();
+            if by_month_u32 <= next_date.month() {
```

Comment:
Not sure about this bit,

so if the current month is 6 and BYMONTH=6, I would still like the next date to fall in this current month and not move a year forward?

Reply:
Good catch! It seems we didn't have any test cases on this edge case. I started refactoring daily and was able to reuse a lot of this same implementation. Due to that, I noticed that this particular conversion fell short. But, I think I have a fix that will work for both daily and monthly. I'll be sure to add new test cases to cover.